### ISSUE TITLE
why is being incorrectly populated in the new→legacy event conversion

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7851]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)| @wsanchez reported|
|-|-|
|Trac ID|trac#7851|
|Type|enhancement|
|Created|2015-04-14 22:56:25Z|
|Branch|https://github.com/twisted/twisted/tree/log-why-7851-2|

`why` is being incorrectly populated in the new→legacy event conversion.

It shouldn't be populated at all.

<details><summary>Searchable metadata</summary>

```
trac-id__7851 7851
type__enhancement enhancement
reporter__wsanchez wsanchez
priority__highest highest
milestone__None None
branch__branches_log_why_7851_2 branches/log-why-7851-2
branch_author__ralphm__wsanchez ralphm, wsanchez
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1429052185715511 1429052185715511
changetime__1445217226370668 1445217226370668
version__None None
owner__ralphm ralphm

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_2"></a>|@wsanchez commented|
|-|-|

(In [44444]) Branch for [#7851](https://github.com/twisted/twisted/issues/7851).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to @wsanchez|
|-|-|

Changes looks good to merge. Don't forget the news file.

Is there any reference in the documentation regarding why "why" should not be populated.

I have check the docstrings from here https://github.com/twisted/twisted/blob/trunk/twisted/logger/_legacy.py and narrative doc from here https://github.com/twisted/twisted/blob/trunk/docs/core/howto/logger.rst

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_6"></a>|@wsanchez commented|
|-|-|

`why` is a not-super-well-documented artifact of the old system, and as a result I misunderstood it's use.

The result was that log output after merging the new system altered the text emitted by logging from clients of the old system (like, say `twistd`), so whoops, this undoes that.  Basically `why` should contain user-provided text, not automatically generated text.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_7"></a>|@wsanchez set status to `closed`|
|-|-|

(In [44465]) Merge branches/log-why-7851: Don't set `why` key in legacy observer wrapper.

Author: wsanchez
Reviewer: adiroiban
Fixes: [#7851](https://github.com/twisted/twisted/issues/7851)

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_8"></a>|@wsanchez commented|
|-|-|

Sigh.  This was the wrong fix, as ralphm found a different case where this results in wrong output.

ralphm has a better test case, and a fix, which requires that this be reverted, so I'm gonna undo the commit to trunk and re-branch.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_9"></a>|@wsanchez set status to `reopened`|
|-|-|

(In [44475]) Revert r44465: Merge branches/log-why-7851

It turns out we need the "why" set set as before, and the cause of
duplicate log output was elsewhere.  We'll re-branch for [[#7851](https://github.com/twisted/twisted/issues/7851)](https://github.com/twisted/twisted/issues/7851) with
another test case and fix.

Reopens: #7851

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_10"></a>|@wsanchez set owner to @ralphm<br>@wsanchez set status to `new`|
|-|-|

To ralphm for fix

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_11"></a>|@ralphm commented|
|-|-|

(In [44477]) Branching to log-why-7851-2.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_12"></a>|@ralphm removed owner|
|-|-|

The original problem we wanted to solve is that in some cases, the output had a duplicate traceback when failures were logged: a brief traceback and a verbose one. This new branch removes the duplicate tracebacks and also adds tests for the log messages. This includes the case where `Deferred.debug` is true, which was previously entirely untested (and actually raised an exception).

Note that the `Deferred.__del__` actually logs two messages. We could fold that into one, but I wasn't sure if that is a backwards incompatible change.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_13"></a>|@wsanchez set owner to @wsanchez<br>@wsanchez set status to `assigned`|
|-|-|

`

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_14"></a>|@wsanchez set owner to @ralphm<br>@wsanchez set status to `new`|
|-|-|

Awesome, thanks, ralphm.

`test_errorLogNoRepr` and `test_errorLogDebugInfo` use an un-numbered format token '{!r}' which needs to be '{0!r}' because python 2.6.  If you fix that and the 2.6 builders are happy, this is good to go.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_15"></a>|@ralphm commented|
|-|-|

(In [44489]) Fix Python 2.6 incompatibility.

Re: [#7851](https://github.com/twisted/twisted/issues/7851).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_16"></a>|@ralphm set status to `closed`|
|-|-|

(In [44491]) Merge log-why-7851-2: Properly log deferred tracebacks.

Author: ralphm
Reviewer: wsanchez
Fixes: [#7851](https://github.com/twisted/twisted/issues/7851)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl commented|
|-|-|

Ticket retargeted after milestone deleted
